support integer-ranges in pow()

staabm opened the pull request:
closes the last remaining todo of #5614

fixes https://github.com/phpstan/phpstan/issues/5614

staabm committed:
support integer-ranges in pow()

staabm commented:
//cc @orklah

ondrejmirtes reviewed (changes requested):
This doesn't handle unions of integer ranges on both sides. Instead of `instanceof *Type`, it'd be better to handle this with a new method on Type. But of course that needs to be designed in order to be useful, ideally to replace other instances of `instancoef IntegerRangeType` and even maybe `TypeUtils::getConstantIntegers()`.

staabm committed:
tests

staabm commented:
> it'd be better to handle this with a new method on Type

if I get you right we are talking about methods such as the following, right?
```diff
diff --git a/src/Type/Type.php b/src/Type/Type.php
index f02971559..5a00d7f1c 100644
--- a/src/Type/Type.php
+++ b/src/Type/Type.php
@@ -112,6 +112,16 @@ interface Type

        public function shuffleArray(): Type;

+       public function subtract(Type $subtrahend): Type;
+
+       public function sum(Type $summand): Type;
+
+       public function multiply(Type $multiplicand): Type;
+
+       public function divide(Type $divisor): Type;
+
+       public function pow(Type $exponent): Type;
+
```

or do you have somthing different in mind?

ondrejmirtes commented:
Yeah, that would be great, although "pow" isn't a word unlike the others, should be "exponentiate" probably.

ondrejmirtes commented:
We'll be able to get rid of a lot of code in InitializerExprTypeResolver :)

staabm committed:
implement Type->exponentiate()

staabm commented:
implemented `Type->exponentiate()` for now and will do the others in separate PRs

staabm commented on src/Reflection/InitializerExprTypeResolver.php:
```diff
@@ -1122,54 +1122,9 @@ public function getPowType(Expr $left, Expr $right, callable $getTypeCallback):
+			return $exponentiatedTyped;
 		}
 
 		return $this->resolveCommonMath(new BinaryOp\Pow($left, $right), $leftType, $rightType);
```
I kept this pass thru `resolveCommonMath` because otherwise we would loose support for operator-type specifiying extensions (even though atm, there is no test-coverage for using it with `pow`).

staabm replied:
test added with https://github.com/phpstan/phpstan-src/pull/1933

ondrejmirtes commented on src/Reflection/InitializerExprTypeResolver.php:
```diff
@@ -1122,54 +1122,9 @@ public function getPowType(Expr $left, Expr $right, callable $getTypeCallback):
+			return $exponentiatedTyped;
 		}
 
 		return $this->resolveCommonMath(new BinaryOp\Pow($left, $right), $leftType, $rightType);
```
What resolveCommonMath does here? Feels like it shouldn't be necessary to call it at all.

staabm replied:
its still required, see https://github.com/phpstan/phpstan-src/pull/1904#discussion_r1007684752

we may drop or at least de-duplicate it, after we moved all the operations into the type-system

ondrejmirtes replied:
So I'd call a different method that only calls those extensions and nothing else.

staabm replied:
I tried and realized there is more shared logic. 

I think it will be simpler after we moved more logic into the type system

staabm replied:
after the final refactoring, I was able to drop the `resolveCommonMath` path

ondrejmirtes commented on src/Type/Accessory/AccessoryNumericStringType.php:
```diff
@@ -216,4 +216,9 @@ public function tryRemove(Type $typeToRemove): ?Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong https://3v4l.org/AoROm

ondrejmirtes commented on src/Type/Accessory/AccessoryNonFalsyStringType.php:
```diff
@@ -204,4 +204,9 @@ public static function __set_state(array $properties): Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong https://3v4l.org/AoROm

ondrejmirtes commented on src/Type/Accessory/AccessoryNonEmptyStringType.php:
```diff
@@ -213,4 +213,9 @@ public function tryRemove(Type $typeToRemove): ?Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong https://3v4l.org/AoROm

ondrejmirtes commented on src/Type/Accessory/AccessoryLiteralStringType.php:
```diff
@@ -204,4 +204,9 @@ public static function __set_state(array $properties): Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong https://3v4l.org/AoROm

ondrejmirtes commented on src/Type/Constant/ConstantIntegerType.php:
```diff
@@ -88,6 +91,40 @@ public function generalize(GeneralizePrecision $precision): Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		if ($exponent instanceof IntegerRangeType) {
```
What about `ConstantIntegerType ** "unions of integer ranges"`? Where is it handled?

ondrejmirtes commented on src/Type/MixedType.php:
```diff
@@ -685,6 +685,14 @@ public function tryRemove(Type $typeToRemove): ?Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new BenevolentUnionType([
```
What if all numeric types are subtracted?

staabm replied:
pow() supports a lot of types as exponent, see e.g. https://3v4l.org/ZtZVc

we would need a big union, which might not be really useful?

per https://www.php.net/manual/en/function.pow.php it supports `mixed` args, while phpstan only accepts `int|float` per functionMap.

ondrejmirtes commented on src/Type/NullType.php:
```diff
@@ -238,6 +238,11 @@ public function getGreaterOrEqualType(): Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong https://3v4l.org/Vvfcu

ondrejmirtes commented on src/Type/ObjectType.php:
```diff
@@ -1320,4 +1320,13 @@ public function tryRemove(Type $typeToRemove): ?Type
+	public function exponentiate(Type $exponent): Type
+	{
+		$object = new ObjectWithoutClassType();
+		if (!$exponent instanceof NeverType && !$object->isSuperTypeOf($this)->no() && !$object->isSuperTypeOf($exponent)->no()) {
```
Objects can be exponentiated? Huh?

staabm replied:
yep, see https://github.com/phpstan/phpstan-src/blob/285b49f2072e18ffe9ff72f221151512b8a92fe6/tests/PHPStan/Analyser/data/pow.php#L17-L25

I moved this pre-exisiting logic from `InitializerExprTypeResolver`

ondrejmirtes commented on src/Type/StaticType.php:
```diff
@@ -562,6 +562,11 @@ public function tryRemove(Type $typeToRemove): ?Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Should call `$this->getStaticObjectType()->exponentiate`

ondrejmirtes commented on src/Type/StringType.php:
```diff
@@ -180,6 +180,11 @@ public function tryRemove(Type $typeToRemove): ?Type
 
+	public function exponentiate(Type $exponent): Type
+	{
+		return new ErrorType();
```
Wrong

staabm committed:
refined non-integer types

staabm committed:
extracted callOperatorTypeSpecifyingExtensions()

staabm committed:
more tests

staabm commented on src/Type/FloatType.php:
```diff
@@ -149,6 +150,33 @@ public function traverse(callable $cb): Type
+		}
+
+		if ($exponent instanceof ConstantScalarType) {
+			if ($exponent->getValue() == 0) {
```
the cs-fixer will turn these into strict comparisons, which in turn breaks a lot of tests.
I think we should ignore these from the cs fixer

staabm commented on src/Type/Constant/ConstantIntegerType.php:
```diff
@@ -88,6 +93,56 @@ public function generalize(GeneralizePrecision $precision): Type
+		} elseif ($exponent instanceof ConstantScalarType) {
+			$exponentValue = $exponent->getValue();
+			if (is_int($exponentValue)) {
+				$min = $this->getValue() ** $exponentValue;
```
Memo to me: double check... seems duplicated with the following if (constant-scalar) case

staabm committed:
refactored logic into ExponentiateHelper

staabm commented:
thanks for the feedback.

after I realized php-src pow() actually supports a lot more types then `int|float`, I refactored this PR and added a lot of more tests.

I could not locale where the last remaining phpstan error is originating from. any hint would be great.
maybe it should just be put into the baseline, since using `null` as a base for exponentiation seems to be something we might want to error about, even if php-src pow() actually allows it?

staabm commented:
Got an idea

staabm committed:
Delete test.php

ondrejmirtes commented:
There are some conflicts.

staabm commented:
resolved

staabm committed:
get rid of unnecessary instanceof

staabm committed:
cheap checks first

ondrejmirtes commented on src/Type/FloatType.php:
```diff
@@ -201,6 +201,12 @@ public function traverse(callable $cb): Type
+	public function exponentiate(Type $exponent): Type
+	{
+		$helper = new ExponentiateHelper();
+		return $helper->exponentiate($this, $exponent);
```
It's weird to `new` this. Typically we inject these services through DI. But I understand that we're in a value object where DI is not accessible. In these cases I'd write this as a static method: `ExponentiateHelper::exponentiate`. Thanks.

staabm replied:
good point, fixed.

ondrejmirtes reviewed (changes requested):
Otherwise 👍

staabm committed:
fix cs

staabm committed:
Update ExponentiateHelper.php

ondrejmirtes commented:
Thank you.
